internal/clients/clientimpl/baseimagematcher/baseimagematcher.go

```diff
@@ -0,0 +1,160 @@
+		g.Go(func() error {
+			if ctx.Err() != nil {
+				return ctx.Err() // this value doesn't matter to errgroup.Wait(), it will be ctx.Err()
+			}
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://api.deps.dev/v3alpha/querycontainerimages/"+chainID.String(), nil)
+			if err != nil {
+				matcher.Reporter.Errorf("failed to build request: %s\n", err)
+				return nil
+			}
+
+			resp, err := matcher.Client.Do(req)
+			if err != nil {
+				matcher.Reporter.Errorf("deps.dev API error: %s\n", err)
+				return nil
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode == http.StatusNotFound {
+				return nil
+			}
+
+			if resp.StatusCode != http.StatusOK {
+				matcher.Reporter.Errorf("deps.dev API error: %s\n", resp.Status)
+				return nil
+			}
+
+			var results struct {
+				Results []struct {
+					Repository string `json:"repository"`
+				} `json:"results"`
+			}
+
+			d := json.NewDecoder(resp.Body)
+			err = d.Decode(&results)
+			if err != nil {
+				matcher.Reporter.Errorf("Unexpected return type from deps.dev base image endpoint: %s", err)
+				return nil
+			}
+
+			// Found some base images!
+			baseImagePossibilities := []models.BaseImageDetails{}
+			for _, r := range results.Results {
+				baseImagePossibilities = append(baseImagePossibilities, models.BaseImageDetails{
+					Name: r.Repository,
+				})
+			}
+			// TODO(v2): Temporary heuristic for what is more popular
+			// Ideally this is done by deps.dev before release
+			slices.SortFunc(baseImagePossibilities, func(a, b models.BaseImageDetails) int {
+				return len(a.Name) - len(b.Name)
+			})
+			baseImagesMap[i] = baseImagePossibilities
+
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, context.DeadlineExceeded
+	}
+
+	allBaseImages := [][]models.BaseImageDetails{
+		// The base image at index 0 is a placeholder representing your image, so always empty
+		// This is the case even if your image is a base image, in that case no layers point to index 0
+		{},
+	}
+
+	currentBaseImageIndex := 0
+	for i, baseImages := range slices.Backward(baseImagesMap) {
+		if len(baseImages) == 0 {
+			layerMetadata[i].BaseImageIndex = currentBaseImageIndex
+			continue
+		}
+
+		// This layer is a base image boundary
+		allBaseImages = append(allBaseImages, baseImages)
+		currentBaseImageIndex += 1
+		layerMetadata[i].BaseImageIndex = currentBaseImageIndex
+
+		// Backfill with heuristic:
+		//   The goal here is to replace empty layers that is currently categorized as the previous base image
+		//   with this base image if it actually belongs to this layer.
+		//
+		//   We do this by guessing the boundary of empty layers by checking for the following commands,
+		//   which are commonly the *last* layer.
+		//
+		//   Remember we are looping backwards in the outer loop,
+		//   so this backfill is actually filling down the layer stack, not up.
+		possibleFinalBaseImageCommands := []string{
+			"/bin/sh -c #(nop)  CMD",
```

Comment:
How reliable is it that we're relying on there always being two? spaces before the `CMD` ?

Reply:
I'm not 100% sure, as this is from the scalibr code. I believe it is relatively reliable, since it's added by build tools, and these two versions of CMD is supporting the current output of `docker build` and `docker buildx build` (aliased to `docker build` after a recent docker version) .

Perhaps this might change in a future release, but at that point we have no idea what the CMD could look like, so I think leaving it as this is fine for now.

Reply:
SGTM